### ISSUE TITLE
RunTargetsAndExit() not exiting

### ISSUE THREAD

**Issue by bonesoul** [adamralph/bullseye#363]
## Version(s)

3.0.0

## To reproduce

```
        public string[] RemainingArguments { get; } = null;

....
            Target("default", DependsOn("build"));

            RunTargetsAndExit(RemainingArguments);
```

just gets stuck and never exiting. any ideas?

**Comment by adamralph** [comment]
@bonesoul are you using `McMaster.Extensions.CommandLineUtils`? If so, it's a [bug in that package](https://github.com/natemcmaster/CommandLineUtils/issues/292).

**Comment by bonesoul** [comment]
seems that's the issue.

**Comment by thomaslevesque** [comment]
@bonesoul in the meantime you can call `RunTargetsWithoutExiting` instead of `RunTargetsAndExit`.

It's not ideal, though, as `RunTargetsWithoutExiting` doesn't handle exceptions. If you don't handle exceptions yourself, they'll be dumped to the console and will crash the process. In contrast, `RunTargetsAndExit` will only dump the message, and will exit with code 1 (in case of a failed target) or 2 (in case of invalid usage).

As far as I can tell, you can't reproduce the exact same behavior by manually handling the exception, because the specific exceptions are private (well, I guess you could just look at their name, but it's a bit hacky...)

@adamralph I wonder if Bullseye could offer another option: `RunTargetsAndReturn`, which would do the same as `RunTargetsAndExit`, but return an exit code instead of calling `Environment.Exit`. I realize it would add a bit of clutter to the API, but it's the only clean solution I can see for this problem.

**Comment by thomaslevesque** [comment]
Given the nature of the problem, I don't expect it to be fixed quickly in `McMaster.Extensions.CommandLineUtils`... It's blocking in the `AppDomain.ProcessExit` handler until execution completes, which will never happen since `Environment.Exit` is blocked until the handler completes... it's a textbook deadlock.

**Comment by adamralph** [comment]
> the specific exceptions are private

That's actually a design flaw and I've raised https://github.com/adamralph/bullseye/issues/367 to fix it. Once that's done, I don't see a need for an exit code returning method, since `RunTargetsWithoutExiting` can be used, and the exceptions can be caught. Thanks for prompting this @thomaslevesque.

Ultimately, https://github.com/natemcmaster/CommandLineUtils/issues/292 needs to be fixed. An app that calls `Environment.Exit` should exit, not hang.

**Comment by thomaslevesque** [comment]
> That's actually a design flaw and I've raised #367 to fix it. Once that's done, I don't see a need for an exit code returning method, since `RunTargetsWithoutExiting` can be used, and the exceptions can be caught.

I see your point, but I think it would be useful nonetheless. Sure, I can catch the exception and convert it to an exit code, but I'd just be repeating what is already done by `RunTargetsAndExit`. And I couldn't even do it exactly the same way, since I don't have access to the logger.

> Ultimately, [natemcmaster/CommandLineUtils#292](https://github.com/natemcmaster/CommandLineUtils/issues/292) needs to be fixed. An app that calls `Environment.Exit` should exit, not hang.

It's true, but I don't see an ideal way to fix it. Either it doesn't handle ProcessExit (and there's a good reason to handle it), or the code can't use `Environment.Exit` without deadlocking. It could block with a timeout, but that would introduce an annoying pause before the program completes.

Also, one could argue that a library shouldn't call `Environment.Exit(0)`. I understand why it makes sense in Bullseye's case, but having an alternative would be nice (ideally, one that doesn't need boilerplate code around it, as is currently the case with `RunTargetsWithoutExiting`).

If you're not strictly opposed to it, I can send a PR to add `RunTargetsAndReturn`, it shouldn't take long.

**Comment by adamralph** [comment]
Thanks for the offer @thomaslevesque but I'd prefer to wait on the outcome of https://github.com/natemcmaster/CommandLineUtils/issues/292 before proceeding down that path.

I'm having a hard time buying the notion of a lib legimately causing an app to hang on a call to `Environment.Exit`.

**Comment by thomaslevesque** [comment]
FYI, https://github.com/natemcmaster/CommandLineUtils/issues/292 is fixed!

**Comment by adamralph** [comment]
Woot! I've confirmed that McMaster.Extensions.CommandLineUtils 2.4.3 fixes the problem.

Kudos @thomaslevesque!

**Comment by bonesoul** [comment]
yep, it's fixed now.
